### PR TITLE
#172: implemented GFW fire alerts map

### PR DESCRIPTION
**Pull request by Yash621**
implemented fire alerts GFW map

screenshots:

image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,7 +12,7 @@ web/modules/*
 !web/themes
 web/themes/*
 !web/themes/custom
-
+forests.sql
```
Comment (pcambra) [inline .gitignore:15]: This change is unrelated

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 2]
```diff
@@ -28,8 +28,12 @@ public function buildForm(array $form, FormStateInterface $form_state) {
     $form['#title'] = $this->t('GFW');
 
     $form['placeholder'] = [
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:30]: Let's rename this 'placeholder' for something more meaningful

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js [hunk 3]
```diff
@@ -1,12 +1,73 @@
...
+    attach: async function (instance) {
       var url = new URL('/nfa-assets/geojson/' + instance.farmMapSettings.plan, window.location.origin + drupalSettings.path.baseUrl)
       var newLayer = instance.addLayer('geojson', {
         title: Drupal.t('Locations'),
         url,
         color: 'orange',
       })
+
+      // Add a layer for fire alerts in the GFW plan tab
+      let gfwRegex = a = /\/plan\/\d+\/gfw/;
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js:12]: Not sure if I understand what this does?

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js [hunk 4]
```diff
@@ -1,12 +1,73 @@
...
+      let windowUrl = window.location.href;
+      if(windowUrl.match(gfwRegex)) {
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js:14]: Please make sure you're following the coding standards for javascript as well

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js (hunk 3 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: @pcambra this is a typo, this must have come while refactoring things out, have reverted this change, apologies will be more careful in the future

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js (hunk 4 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: surely

File: .gitignore (hunk 1 above)
Comment (Yash621) [reply .gitignore]: reverted

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php (hunk 2 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php]: @pcambra  have renamed it to `gfw_map_placeholder` will that do, or we can have something more better for this ?

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 5]
```diff
@@ -27,9 +27,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
     // Set the form title.
     $form['#title'] = $this->t('GFW');
 
-    $form['placeholder'] = [
-      '#type' => 'markup',
-      '#markup' => '(GFW form placeholder)',
+    $form['gfw_map_placeholder'] = [
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:30]: A placeholder is a text that is there until is replaced by something else, we've now replaced it by a map, so "map" or "gfw_map" makes more sense

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js [hunk 6]
```diff
@@ -1,12 +1,73 @@
...
+              }
+              instance.addLayer('geojson', {
+                title: 'Fire Alerts',
+                geojson : geoJson,
+                color: 'red'
+              });
+            }
+          })
+        } catch(err) {
+          console.log(err);
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js:66]: Let's not leave console traces as a general advice

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 7]
```diff
@@ -27,9 +27,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+      '#type' => 'farm_map',
+      '#map_type' => 'farm_nfa_plan_locations',
+      '#map_settings' => [
+        'plan' => \Drupal::routeMatch()->getRawParameter('plan')
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:34]: Please check what "dependency injection" is and inject the route match instead, it's a better practice

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js [hunk 8]
```diff
@@ -1,12 +1,73 @@
...
+    attach: async function (instance) {
       var url = new URL('/nfa-assets/geojson/' + instance.farmMapSettings.plan, window.location.origin + drupalSettings.path.baseUrl)
       var newLayer = instance.addLayer('geojson', {
         title: Drupal.t('Locations'),
         url,
         color: 'orange',
       })
+
+      // Add a layer for fire alerts in the GFW plan tab
+      let gfwRegex = /\/plan\/\d+\/gfw/;
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js:12]: I still don't clearly understand what this does? could you walk me through it on slack? thanks!

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php (hunk 5 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php]: got it !

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js (hunk 6 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: got it, will keep that in mind, thanks !

File: web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js (hunk 8 above)
Comment (Yash621) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: surely, to brief though what this is apparently doing is the map is basically been rendered at many routes but we don't want
the fire and deforestation layer to get added on every route, we want it to get added just on the GFW section, so basically to check that i am using a regex that , matches the GFW route and runs the code to add the fires and deforestation layer just for the GFW section. hope this clears up this a bit
happy to sync on slack too to discuss this further !

Comment (pcambra) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: @Yash621 please check how to do this directly on PHP rather than js, for example https://www.computerminds.co.uk/articles/context-aware-blocks

Comment (Yash621) [reply web/modules/custom/farm_nfa/js/farmOS.map.behaviors.farm_nfa_plan_locations_layers.js]: hey @pcambra instead of context-aware-blocks, I am thinking of passing an attribute like `'id' => 'gfw-map-component'` in the `forestPlanGfwForm.php` and checking the existence of the `gfw map` based on that in js file, will this approch be viable, or will using contextual blocks be a better option instead ?

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 9]
```diff
@@ -11,6 +11,12 @@
  * @ingroup farm_nfa
  */
 class ForestPlanGfwForm extends FormBase {
+  
+  protected $routeMatch;
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:15]: Please check how other class variables are declared thorough the Drupal standards. Your IDE should be warning you about this, you can configure VSCode or PHPStorm with https://www.drupal.org/docs/develop/standards, see https://www.drupal.org/project/coder

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 10]
```diff
@@ -11,6 +11,12 @@
...
+
+  public function __construct() {
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:17]: This is not how traditionally DI works, please look for examples in the existing code and the books :)

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 11]
```diff
@@ -27,9 +33,17 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
     $form['#title'] = $this->t('GFW');
 
-    $form['placeholder'] = [
-      '#type' => 'markup',
-      '#markup' => '(GFW form placeholder)',
+    $form['gfw_map'] = [
+      '#type' => 'farm_map',
+      '#map_type' => 'farm_nfa_plan_locations',
+      '#map_settings' => [
+        'plan' => $this->routeMatch->getRawParameter('plan')
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:40]: You're quite right here, I was giving you examples of blocks which are not accurate, apologies.

You should be getting the plan by using getParameter, that should be of PlanInterface, you can check if it is before rendering anything and avoid checking regex in Javascript.

You can also pass drupalSettings I think to the library, see https://www.drupal.org/docs/drupal-apis/javascript-api/javascript-api-overview

Comment (Yash621) [reply web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:65]: no worries @pcambra , the regex check in Javascript has been already been removed as i am attacahing the library separately now in the form, so we no longer need to do any checks in javascript. though as suggested after using `getParameter` instead of `getRawParameter` i am getting an empty object `{}` for `plan` parameter  instead of the expected plan id that i was getting on using the `getRawParameter`

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 12]
```diff
@@ -4,13 +4,44 @@
...
+   */
+  public function __construct(RouteMatchInterface $routeMatch) {
+    $this->routeMatch = $routeMatch;
+  }
+  
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    // Instantiates this form class.
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:38]: no need for this comment

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 13]
```diff
@@ -4,13 +4,44 @@
...
+    return new static(
+      // Load the service required to construct this class.
+      $container->get('current_route_match')
+      //
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:42]: or this

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php [hunk 14]
```diff
@@ -4,13 +4,44 @@
...
+    $this->routeMatch = $routeMatch;
+  }
+  
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    // Instantiates this form class.
+    return new static(
+      // Load the service required to construct this class.
```
Comment (pcambra) [inline web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:40]: or this :)

File: web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php (hunk 11 above)
Comment (pcambra) [reply web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:65]: We can live with getRawParameter for the time being but probably need to look at why getParameter is not working.

You're still using:

```
  let planId = window.location.href.split('/')[4];
  let cfrPlanUrl = `${window.origin}/nfa-assets/geojson/${planId}`;
```

When you can pass the plan id along the URL as drupalSetting instead.

Comment (Yash621) [reply web/modules/custom/farm_nfa/src/Form/ForestPlanGfwForm.php:62]: @pcambra when you said to pass the plan id along the URL as drupalSetting instead, do you mean to pass only the planId
as a drupal setting and then use it in the cfrPlanUrl ? or pass both the planId and cfrPlanUrl as a drupal setting instead ?
